src/devices/bus/msx/ctrl/ctrl.h

```diff
@@ -0,0 +1,85 @@
+// license:BSD-3-Clause
+// copyright-holders:Wilbert Pol
+/**********************************************************************
+
+    MSX General Purpose port emulation
+
+The MSX General Purpose port uses a DE-9 connector; the pin assignment
+is similar to but slightly different from the Atari VCS controller port.
+
+**********************************************************************/
+
+#ifndef MAME_BUS_MSX_CTRL_CTRL_H
+#define MAME_BUS_MSX_CTRL_CTRL_H
+
+#pragma once
+
+
+class msx_general_purpose_port_device;
+
+
+class device_msx_general_purpose_port_interface : public device_interface
+{
+public:
+	virtual ~device_msx_general_purpose_port_interface() { }
+
+	virtual u8 read() { return 0xff; }
+	virtual void pin_6_w(int state) { }
+	virtual void pin_7_w(int state) { }
+	virtual void pin_8_w(int state) { }
+
+protected:
+	device_msx_general_purpose_port_interface(const machine_config &mconfig, device_t &device);
+
+	msx_general_purpose_port_device *m_port;
+};
+
+
+class msx_general_purpose_port_device : public device_t,
+								public device_slot_interface
```

Comment:
If you use `device_single_card_slot_interface<device_msx_general_purpose_port_interface>` as the base, you get validation and downcasts for free.

Reply:
Thanks, changed the base to device_single_card_slot_interface.